walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,133 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (events.isEmpty() && request.getDuration() <= TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    
+    List<TimeRange> availableTimes = new ArrayList<TimeRange>();
+    List<Event> eventList = new ArrayList<>(events);
+
+    Collections.sort(eventList, Event.ORDER_BY_START);
+
+    Event firstEvent = eventList.get(0);
+    Event lastEvent = eventList.get(eventList.size() - 1);
+    int eventsSkipped = 0;
+    boolean requestHasEventAttendees = requestHasEventAttendees(request, firstEvent);
+
+    // If there is only one event happening that day, the for loop will be skipped,
+    // so we handle the case separately.
+    if (eventList.size() == 1) {
+      if (!requestHasEventAttendees) {
+        eventsSkipped++;
+      } 
+    }
+
+    // Add the first time slot, from the start of the day to the start of the first event,
+    // only if there's no event that starts the day.
+    if (startOfDayIsFree(eventList) && requestHasEventAttendees) {
+      availableTimes.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY,
+        firstEvent.getWhen().start(), false));
+    }
+
+    for (int i = 0; i < eventList.size() - 1; i++) {
+      TimeRange currEventTimeRange = eventList.get(i).getWhen();
+      TimeRange nextEventTimeRange = eventList.get(i + 1).getWhen();
+
+      if (!requestHasEventAttendees(request, eventList.get(i))) {
+        eventsSkipped ++;
+        continue;
+      }
+
+      if (currEventTimeRange.contains(nextEventTimeRange)) {
+        // If there is a 3rd event, connect events 1 and 3. Otherwise, connect with the end of day.
+        if (i + 2 < eventList.size() - 1) {
+          TimeRange nextnextEventTimeRange = eventList.get(i + 2).getWhen();
+          availableTimes.add(TimeRange.fromStartEnd(currEventTimeRange.end(),
+            nextnextEventTimeRange.start(), true));
+          i++;
+        } else {
+          availableTimes.add(TimeRange.fromStartEnd(currEventTimeRange.end(),
+            TimeRange.END_OF_DAY, true));
+        }
+      } else {
+          // If current event completely overlaps the next event, there's no reason to check
+          // the next one too, as it's start and end are completely contained.
+          if (currEventTimeRange.overlaps(nextEventTimeRange)) {
+            continue;
+          }
+
+          if (nextEventTimeRange.start() - currEventTimeRange.end() >= request.getDuration()) {
+            availableTimes.add(TimeRange.fromStartEnd(currEventTimeRange.end(),
+              nextEventTimeRange.start(), false));  
+          } 
```

Comment:
I think this also depends on the attendee list for the next event. If none of the requested attendees are occupied by the next meeting, then the range does not need to stop at the start of the next meeting

Reply:
I think both can be solved if I just keep a list with events that have attendees, instead of going through all of them

Reply:
That would definitely make it easier if you first filtered out all the meetings that don't have any relevant attendee